Dalamud/Game/Gui/Nameplates/NameplateGui.cs

```diff
@@ -0,0 +1,212 @@
+    }
+
+    private unsafe delegate nint SetPlayerNameplateDetourDelegate(nint playerNameplateObjectPtr, bool isTitleAboveName, bool isTitleVisible, nint titlePtr, nint namePtr, nint freeCompanyPtr, nint prefix, int iconId);
+
+    /// <inheritdoc/>
+    public event INameplatesGui.OnNameplateUpdateDelegate OnNameplateUpdate;
+
+    /// <inheritdoc/>
+    public void DisposeService()
+    {
+        this.setPlayerNameplateDetourHook.Dispose();
+    }
+
+    /// <summary>
+    /// Gets the index of a given nameplate.
+    /// </summary>
+    /// <param name="nameplateObjectPtr">The nameplate object pointer where you want the index from.</param>
+    /// <returns>Returns the index for the given nameplate.</returns>
+    public long GetNameplateIndex(nint nameplateObjectPtr)
+    {
+        // Get the nameplate object array
+        var nameplateObjectArrayPtrPtr = this.namePlatePtr + Marshal.OffsetOf(typeof(AddonNamePlate), nameof(AddonNamePlate.NamePlateObjectArray)).ToInt32();
+        var nameplateObjectArrayPtr = Marshal.ReadIntPtr(nameplateObjectArrayPtrPtr);
+
+        if (nameplateObjectArrayPtr == nint.Zero)
+            return -1;
+
+        // Determine the index of the nameplate object within the nameplate object array
+        var namePlateObjectSize = Marshal.SizeOf(typeof(AddonNamePlate.NamePlateObject));
+        var namePlateObjectPtr0 = nameplateObjectArrayPtr + namePlateObjectSize * 0;
+        var namePlateIndex = (nameplateObjectPtr.ToInt64() - namePlateObjectPtr0.ToInt64()) / namePlateObjectSize;
+
+        if (namePlateIndex < 0 || namePlateIndex >= AddonNamePlate.NumNamePlateObjects)
+            return -1;
+
+        return namePlateIndex;
+    }
+
+    /// <summary>
+    /// Gets the corresponding <see cref="GameObject"/> for the nameplate by its index.
+    /// </summary>
+    /// <param name="namePlateIndex">The index of the namplate where you want to get the <see cref="GameObject"/> for.</param>
+    /// <returns>Returns the corresponding <see cref="GameObject"/> for the nameplate.</returns>
+    public GameObject? GetNameplateGameObject(long namePlateIndex)
+    {
+        if (namePlateIndex == -1)
+            return null;
+
+        // Get the nameplate info
+        RaptureAtkModule.NamePlateInfo namePlateInfo;
+        unsafe
+        {
+            var framework = FFXIVClientStructs.FFXIV.Client.System.Framework.Framework.Instance();
+            namePlateInfo = framework->GetUIModule()->GetRaptureAtkModule()->NamePlateInfoEntries[(int)namePlateIndex];
+        }
+
+        // Return the object for its object id
+        var objectId = namePlateInfo.ObjectId.ObjectId;
+        return this.objectTable.SearchById(objectId);
+    }
+
+    private void AddonLifecycle_Event(AddonEvent type, AddonArgs args)
+    {
+        if (type == AddonEvent.PostSetup)
+            this.namePlatePtr = args.Addon;
+        else if (type == AddonEvent.PreFinalize)
+            this.namePlatePtr = nint.Zero;
+    }
```

Comment:
I strongly suggest using two different function for Setup and Finalize, it'll make intent more clear and help prevent bugs like what was fixed earlier.

However this is just a style suggestion, feel free to disregard if you prefer.